### PR TITLE
Auto-pause repeaters

### PR DESCRIPTION
**Pull request by kaapstorm**
Jira: [SC-756](https://dimagi-dev.atlassian.net/browse/SC-756)

##### SUMMARY
If a repeater has attempted to deliver payloads, and has always failed over a period of three months, then it will automatically be paused.

If it has a streak of 10,000 failures, it will also be paused.

(Time starts when this is deployed. It does not analyse historical behavior.)

##### FEATURE FLAG
None (Pro plan or higher)

##### RISK ASSESSMENT / QA PLAN
I would prefer to use unit tests to verify this functionality. We could attempt to send 10,001 failed payloads, but waiting 3 months seems an impractical way to check that this change pauses bad repeaters.

It may be valuable to check that this change does not pause good repeaters! But I am hoping that reviewers can flag **if they think these unit tests are not enough** to verify that.

##### PRODUCT DESCRIPTION
Automatically pause data forwarding if it has always failed throughout a period of three months.

### CODE REVIEW

**Comment by kaapstorm** [comment]
Question about the "[Fix tests](https://github.com/dimagi/commcare-hq/pull/28276/commits/e6a3094469f89fb29c0de46463587f39b23d6429)" commit: I can't think of places in HQ where we might get ConflictErrors because of the change in this PR, but I am worried that the problem that the `reloaded()` function avoids is something we could come across.

Is that a legitimate concern?

And if so, how should we address it? Can we rely on Travis to find all instances of the problem? Should this go through QA? Can we rely on QA to find all instances of the problem?

File: corehq/motech/repeaters/models.py [hunk 1]
```diff
@@ -273,6 +273,28 @@ def update_failure_streak(self, attempt):
             self.failure_streak += 1
         self.save()
 
+    def is_connection_working(self):
+        """
+        Tries to determine whether the remote API is accepting payloads
+        based on past success.
+        """
+        three_months = timedelta(days=90)
+        if self.failure_streak > 10_000:
```
Comment (esoergel) [inline corehq/motech/repeaters/models.py:282]: huh, I've never seen that before, pretty cool!

File: corehq/motech/repeaters/models.py [hunk 2]
```diff
@@ -273,6 +273,28 @@ def update_failure_streak(self, attempt):
...
+            return True
+        if self.last_success_at is None:
+            if self.failure_streak == 0:
+                # Never succeeded, but never failed either: Nothing sent yet.
+                return True
+        else:
+            if datetime.utcnow() - self.last_success_at < three_months:
+                # Has succeeded at least once in the last 3 months
+                return True
+        return False
```
Comment (esoergel) [inline corehq/motech/repeaters/models.py:296]: nit: it's hard to reason about what would result in this final `False` being hit - what would you think of this?
```suggestion
        if self.last_success_at is None:
            # Never succeeded, but never failed either: Nothing sent yet.
            return self.failure_streak == 0:
        else:
            # Has succeeded at least once in the last 3 months
            return datetime.utcnow() - self.last_success_at < three_months:
```

**Review by esoergel — APPROVED** [review]
This does seem reasonably safe to me.  One thing to consider is whether there's a middle ground approach for the 10k limit.  Imagine a high volume repeater that hits a temporary outage - could it reach that 10k limit in a matter of hours?  Then even if it's a temporary outage that's quickly resolved, they'd have to manually un-pause the repeater.

Could we instead rate limit requests until we see a success?  For instance, say we send 5 records a second, but there's a 100% failure rate.  Maybe we only send a maximum of one request a minute until one suceeds, and flag the rest for retry as if they'd failed.  That sounds way harder to implement, so I imagine we'd punt on it.  Or should it factor in the length of time since the last success in some way?  I think what you have here is a decent balance between simplicity and utility, so it's fine by me.

File: corehq/motech/repeaters/models.py (hunk 2 above)
Comment (kaapstorm) [reply corehq/motech/repeaters/models.py:296]: Oh that's much nicer!

**Comment by kaapstorm** [comment]
I like the idea of rate limiting retry attempts. (The first attempt is always sent immediately, but it's the retries that clog up the "queue" (not a real queue).) I'll see if there is a relatively simple way to implement it.

File: corehq/motech/repeaters/models.py [hunk 3]
```diff
@@ -262,6 +265,33 @@ def allowed_to_forward(self, payload):
...
+        three_months = timedelta(days=90)
+        if self.failure_streak > 10_000:
+            # 10,000 misses without a single hit.
+            return False
+        if datetime.utcnow() - self.started_at < three_months:
+            # Three months grace period to get it working
+            return True
+        if self.last_success_at is None:
+            # Never succeeded, but never failed either: Nothing sent yet.
+            return self.failure_streak == 0:
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/motech/repeaters/models.py]: oops

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]
:star:

**Pull request merged**